File: 041/FirstMissingPositive.scala
object FirstMissingPositive {
  def firstMissingPositive(nums: Array[Int]): Int = {
    for (i <- nums.indices) {
      while (0 <= nums(i)-1 && nums(i)-1 < nums.length && nums(i) != nums(nums(i)-1)) {
        val temp = nums(i)
        nums(i) = nums(temp - 1)
        nums(temp - 1) = temp
      }
    }

    for (i <- nums.indices) {
      if (nums(i) != i+1) {
        return i+1
      }
    }
    return nums.length + 1
  }

  def main(args: Array[String]): Unit = {
    val nums = Array(1, 2, 4, 5)
    println(firstMissingPositive(nums))
  }
}

File: 041/first-missing-positive.go
package main

import "fmt"

func firstMissingPositive(nums []int) int {
	for i := 0; i < len(nums); i++ {
		for 0 <= nums[i]-1 && nums[i]-1 < len(nums) && nums[i] != nums[nums[i]-1] {
			nums[i], nums[nums[i]-1] = nums[nums[i]-1], nums[i]
		}
	}

	for k := range nums {
		if nums[k] != k+1 {
			return k + 1
		}
	}
	return len(nums) + 1
}

func main() {
	nums := []int{7, 8, 9, 11, 12}
	fmt.Println(firstMissingPositive(nums))
}
